### PR TITLE
 Check that all tree nodes are unique.

### PR DESCRIPTION
**Pull request by sandreenko**
Check that all tree nodes have unique gtTreeId.

The first commit refactors existing functions.
The second adds the checker and shows the problem.
The third refactors `CopyFrom` function.

### CODE REVIEW

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -19314,3 +19314,35 @@ CORINFO_RESOLVED_TOKEN* Compiler::impAllocateToken(CORINFO_RESOLVED_TOKEN token)
...
+    while (args != nullptr)
+    {
+        GenTreePtr arg = args->Current();
+        if (arg->OperGet() == GT_RET_EXPR)
+        {
+            storeRetExprAsLocalVar(args->pCurrent());
+        }
+        else if ((arg->OperGet() == GT_CAST) && (arg->gtFlags & GTF_CALL))
+        {
+            // Ret_expr can be a normalized small type value returned by an unmanaged function.
```
Comment (sandreenko) [inline src/jit/importer.cpp]: I am not sure that this situation is possible.

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86 full_opt legacy_backend CoreCLR Perf Tests Correctness

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/jit/compiler.hpp [hunk 2]
```diff
@@ -1456,20 +1456,38 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
...
+    // To satisfy SDSU we can copy only isolated nodes, that are not in the blocks.
+    // If the source is not isolated - create a clone.
+    if ((src->gtNext != nullptr) || (src->gtPrev != nullptr))
+    {
+        src = comp->gtCloneExpr(src);
+    }
+
+    assert((src->gtNext == nullptr) && (src->gtNext == nullptr));
+
+    // Mark that after the copying source is built-in the blocks.
```
Comment (CarolEidt) [inline src/jit/compiler.hpp]: I'd suggest rewording this as "Copy the execution order links to 'src', so that when we copy it to 'this', below, those links remain correct."

File: src/jit/compiler.hpp [hunk 3]
```diff
@@ -1456,20 +1456,38 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
...
+//    src - source tree, that replaces this.
+//    comp - the compiler instanse to transfer annotations for arrays.
+//
+inline void GenTree::CopyFrom(GenTree* src, Compiler* comp)
 {
     /* The source may be big only if the target is also a big node */
 
     assert((gtDebugFlags & GTF_DEBUG_NODE_LARGE) || GenTree::s_gtNodeSizes[src->gtOper] == TREE_NODE_SZ_SMALL);
-    GenTreePtr prev = gtPrev;
-    GenTreePtr next = gtNext;
```
Comment (CarolEidt) [inline src/jit/compiler.hpp]: I would suggest retaining this, and clearing the next/prev on 'this' when/if you clone it. Then you avoid the situation where there's some remaining link to 'src' in the tree, with those fields set to those old values.

File: src/jit/flowgraph.cpp [hunk 4]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
     }
 }
 
+//------------------------------------------------------------------------------
+// fgDebugCheckNodesUniqueness: Check that each tree in the method has its own unique gtTreeId.
+//
```
Comment (CarolEidt) [inline src/jit/flowgraph.cpp]: The cost of this method seems prohibitive, even for a checked build. It would seem that an approach that retains ranges of IDs that have been seen (e.g. using a B/B+ tree) would be preferable.

File: src/jit/compiler.hpp [hunk 5]
```diff
@@ -1456,20 +1456,38 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
     }
 }
 
-inline void GenTree::CopyFrom(const GenTree* src, Compiler* comp)
+//------------------------------------------------------------------------
+// CopyFrom: replace this with the src node. If the source is isolated node, then use it as it.
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.hpp]: I thought about that, in that case I will rename function to `ReplaceWith`.

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: This algorithm visit each node only once and it is linear in terms of trees.

File: src/jit/compiler.hpp [hunk 6]
```diff
@@ -1456,20 +1456,38 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
...
-    GenTreePtr next = gtNext;
+
+    // To satisfy SDSU we can copy only isolated nodes, that are not in the blocks.
+    // If the source is not isolated - create a clone.
+    if ((src->gtNext != nullptr) || (src->gtPrev != nullptr))
+    {
+        src = comp->gtCloneExpr(src);
+    }
+
+    assert((src->gtNext == nullptr) && (src->gtNext == nullptr));
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.hpp]: Thanks

File: src/jit/compiler.hpp [hunk 7]
```diff
@@ -1456,20 +1456,38 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
...
 }
 
-inline void GenTree::CopyFrom(const GenTree* src, Compiler* comp)
+//------------------------------------------------------------------------
+// CopyFrom: replace this with the src node. If the source is isolated node, then use it as it.
+//           Otherwise, make a src clone.
+//
+// Arguments:
+//    src - source tree, that replaces this.
+//    comp - the compiler instanse to transfer annotations for arrays.
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.hpp]: Thanks

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (CarolEidt) [reply src/jit/flowgraph.cpp]: > This algorithm visits each node only once and it is linear in terms of trees.

Perhaps the bit-vector is not too prohibitive, given that it is linear cost. However, the algorithm, as far as I can tell, pushes *all* the nodes onto a stack and never pops them.
I don't think you need a stack at all for the LIR case - just do the check & add as you encounter each node. And in the non-LIR case you should pop & check & add the entire stack after each statement is handled.

Comment (sandreenko) [reply src/jit/flowgraph.cpp]: > However, the algorithm, as far as I can tell, pushes all the nodes onto a stack and never pops them

You are right, it is suffering from lack of std collections.
I can't pass bitVector to the visitor, without creating a decorator, because there is only one argument for the data.

Comment (mikedn) [reply src/jit/flowgraph.cpp]: > because there is only one argument for the data.

I don't follow, what do you need more than one argument for?

Comment (sandreenko) [reply src/jit/flowgraph.cpp]: `BitVecOps::AddElemD(&nodesVecTraits, uniqueNodes, gtTreeID); `
we need the bit vector and its traits. 

I pushed a compromise, where we add trees for for one block each time, then check.

File: src/jit/compiler.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.hpp]: When our nodes are not linked these gtPrev and gtNext checks are not valuable. Do we have a way to express that src node is in incorrect state and should not be used? Like set gtTreeId to -1 or smth like that.

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (mikedn) [reply src/jit/flowgraph.cpp]: Ah, traits, yes.

But I still don't understand why all this "complication" with the stack. Make a local struct/class. Put the bit vector and its traits in it. Pass its address to `fgWalkTreePre`. Done.

And you can put `fgMarkTreeId` in it to avoid adding more functions to `Compiler`. There are already way too many.

File: src/jit/flowgraph.cpp [hunk 8]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
...
+void Compiler::fgDebugCheckNodesUniqueness()
+{
+    unsigned     numberOfTrees = JitTls::GetCompiler()->compGenTreeID;
+    BitVecTraits nodesVecTraits(numberOfTrees, this);
+    BitVec       uniqueNodes(BitVecOps::MakeEmpty(&nodesVecTraits));
+
+    ArrayStack<unsigned> stack(this);
+
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: I think the recommendation is to use `block != nullptr` rather than just `block`.

File: src/jit/flowgraph.cpp [hunk 9]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
...
+    {
+        if (block->IsLIR())
+        {
+            for (auto i : LIR::AsRange(block))
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: This looks like an unwarranted use of `auto`, `GenTree*` is preferable.

File: src/jit/flowgraph.cpp [hunk 10]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
...
+            {
+                stack.Push(i->gtTreeID);
+            }
+        }
+        else
+        {
+            for (GenTreeStmt* stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: `stmt != nullptr`

File: src/jit/flowgraph.cpp [hunk 11]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
     }
 }
 
+//------------------------------------------------------------------------------
+// fgDebugCheckNodesUniqueness: Check that each tree in the method has its own unique gtTreeId.
+//
+void Compiler::fgDebugCheckNodesUniqueness()
+{
+    unsigned     numberOfTrees = JitTls::GetCompiler()->compGenTreeID;
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: `JitTls::GetCompiler()` seems odd, this function is a member of `Compiler`.

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (CarolEidt) [reply src/jit/flowgraph.cpp]: > Make a local struct/class. Put the bit vector and its traits in it.

Yes, that's what most of the tree walkers do.

File: src/jit/flowgraph.cpp [hunk 12]
```diff
@@ -21721,6 +21742,58 @@ void Compiler::fgDebugCheckBlockLinks()
...
+//------------------------------------------------------------------------
+// fgMarkTreeId: Visit all subtrees in the tree and memorize gtTreeIDs.
+//
+// Arguments:
+//    pTree            - Pointer to the tree to walk
+//    fgWalkPre        - ArrayStack to memorize gtTreeID
+//
+Compiler::fgWalkResult Compiler::fgMarkTreeId(GenTree** pTree, fgWalkData* fgWalkPre)
+{
+    ArrayStack<unsigned>* stack = reinterpret_cast<ArrayStack<unsigned>*>(fgWalkPre->pCallbackData);
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: `pCallbackData` is `void*` so `static_cast` is enough

File: src/jit/flowgraph.cpp [hunk 13]
```diff
@@ -21597,70 +21597,91 @@ void Compiler::fgDebugCheckLinks(bool morphTrees)
...
+void Compiler::fgDebugCheckStmtsList(BasicBlock* block, bool morphTrees)
+{
+    for (GenTreeStmt* stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
+    {
+        /* Verify that bbTreeList is threaded correctly */
+        /* Note that for the GT_STMT list, the gtPrev list is circular. The gtNext list is not: gtNext of the
+        * last GT_STMT in a block is nullptr. */
 
-                /* For each statement check that the exception flags are properly set */
+        noway_assert(stmt->gtPrev);
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: Isn't `assert` enough?

Comment (sandreenko) [reply src/jit/flowgraph.cpp]: For debug functions there is no difference between assert and noway_assert, but historically there were noway_assert there. Does anybody know why?

File: src/jit/flowgraph.cpp (hunk 12 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed, I will create a PR to fix other places where we use `reinterpret_cast` in the walkers.

File: src/jit/flowgraph.cpp (hunk 11 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed

File: src/jit/flowgraph.cpp (hunk 10 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed

File: src/jit/flowgraph.cpp (hunk 9 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed

File: src/jit/flowgraph.cpp (hunk 8 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed

File: src/jit/flowgraph.cpp (hunk 13 above)
Comment (briansull) [reply src/jit/flowgraph.cpp]: Someone incorrectly change it to a noway_assert.  It doesn't make sense to have a noway_assert in DEBUG only code.

Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Thanks, I will change them back in another PR.

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed, thanks. It looks much better now.

File: src/jit/flowgraph.cpp [hunk 14]
```diff
@@ -21742,58 +21742,76 @@ void Compiler::fgDebugCheckBlockLinks()
     }
 }
 
+// UniquenessCheckWalker keeps data that is neccesary to check
+// that each tree has it is own unique id and they do not repeat.
+class UniquenessCheckWalker
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: I don't know what others think but I would have preferred a local class without all the standard documentation clutter. This looks like overkill just to set a bit in a bitvector. Maybe that's just me :)

Comment (sandreenko) [reply src/jit/flowgraph.cpp]: I do not like to write them, but when they are already written I am fine with them)
I should ask @BruceForstall's opinion when he comes back from his vacation.

File: src/jit/compiler.hpp (hunk 3 above)
Comment (sandreenko) [reply src/jit/compiler.hpp]: Ok, I will create a fix for such places in two separate PRs, because there are many changes there.
This PR will wait until I merge fixes.

File: src/jit/compiler.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/compiler.hpp]: > Do we have a way to express that src node is in incorrect state and should not be used? Like set gtTreeId to -1 or smth like that.

I believe that setting the `gtOper` to `GT_NONE` might be a reasonable approach.

File: src/jit/compiler.hpp (hunk 3 above)
Comment (briansull) [reply src/jit/compiler.hpp]: There is also DEBUG_DESTROY_NODE which is the recommended way to mark a node that should no longer be referenced in the graph of nodes.

File: src/jit/compiler.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/compiler.hpp]: There is also DEBUG_DESTROY_NODE which is the recommended way to mark a node that should no longer be referenced in the graph of nodes.

[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/compiler.hpp]: FYI:

/*****************************************************************************
 * This node should not be referenced by anyone now. Set its values to garbage
 * to catch extra references
 */

inline void DEBUG_DESTROY_NODE(GenTreePtr tree)
{
#ifdef DEBUG
    // printf("DEBUG_DESTROY_NODE for [0x%08x]\n", tree);

    // Save gtOper in case we want to find out what this node was
    tree->gtOperSave = tree->gtOper;

    tree->gtType = TYP_UNDEF;
    tree->gtFlags |= 0xFFFFFFFF & ~GTF_NODE_MASK;
    if (tree->OperIsSimple())
    {
        tree->gtOp.gtOp1 = tree->gtOp.gtOp2 = nullptr;
    }
    // Must do this last, because the "gtOp" check above will fail otherwise.
    // Don't call SetOper, because GT_COUNT is not a valid value
    tree->gtOper = GT_COUNT;
#endif
}

[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.hpp]: Thanks, it is very helpful.

[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/compiler.hpp]: > There is also DEBUG_DESTROY_NODE which is the recommended way to mark a node that should no longer be referenced in the graph of nodes.

Thanks, Brian - I'd forgotten about that!

**Comment by sandreenko** [comment]
@dotnet-bot test Perf Build and Test
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by sandreenko** [comment]
The PR was force updated, the comments were fixed. Please take another look.
